community/proc/src/main/java/org/neo4j/proc/ProcedureRegistry.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.proc;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.neo4j.kernel.api.exceptions.ProcedureException;
+import org.neo4j.kernel.api.exceptions.Status;
+
+public class ProcedureRegistry
+{
+    private final Map<ProcedureSignature.ProcedureName,Procedure> procedures = new HashMap<>();
+
+    /**
+     * Register a new procedure. This method must not be called concurrently with {@link #get(ProcedureSignature.ProcedureName)}.
+     * @param proc the procedure.
+     */
+    public synchronized void register( Procedure proc ) throws ProcedureException
+    {
+        ProcedureSignature signature = proc.signature();
+        ProcedureSignature.ProcedureName name = signature.name();
+
+        validateSignature( signature, signature.inputSignature(), "input" );
+        validateSignature( signature, signature.outputSignature(), "output" );
+
+        if( procedures.putIfAbsent( name, proc ) != null )
+        {
+            throw new ProcedureException( Status.Procedure.FailedRegistration, "Unable to register procedure, because the name `%s` is already in use.", name );
+        }
+    }
+
+    private void validateSignature( ProcedureSignature signature, List<ProcedureSignature.FieldSignature> fields, String fieldType )
+            throws ProcedureException
+    {
+        Set<String> names = new HashSet<>();
+        for ( ProcedureSignature.FieldSignature field : fields )
+        {
+            if( !names.add( field.name() ) )
+            {
+                throw new ProcedureException(  Status.Procedure.FailedRegistration,
+                        "Procedure `%s` cannot be registered, because it contains a duplicated "+fieldType+" field, '%s'. " +
+                        "You need to rename or remove one of the input fields.", signature.toString(), field.name() );
```

Comment:
when a ProcedureException is thrown, it will always say "You need to rename or remove one of the input fields.". Change "input" to "duplicate" in this case, as it could be an output field as well.